File: src/data/lineageEdgesService.ts
import {
  collection,
  doc,
  getDocsFromCache,
  onSnapshot,
  query,
  serverTimestamp,
  setDoc,
  where,
} from '@react-native-firebase/firestore';
import { db, isFirebaseConfigured } from '../firebase/firebase';
import { trackPendingWrite } from '../sync/pendingWrites';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ItemLineageMovementKind = 'sold' | 'returned' | 'correction' | 'association';
export type ItemLineageSource = 'app' | 'server' | 'migration';

export type ItemLineageEdge = {
  id: string;
  accountId: string;
  itemId: string;
  fromTransactionId: string | null;
  toTransactionId: string | null;
  movementKind: ItemLineageMovementKind;
  source: ItemLineageSource;
  createdAt: unknown;
  createdBy?: string | null;
  note?: string | null;
  fromProjectId?: string | null;
  toProjectId?: string | null;
};

// ---------------------------------------------------------------------------
// Normalizer
// ---------------------------------------------------------------------------

export function normalizeEdgeFromFirestore(raw: unknown, id: string): ItemLineageEdge {
  const data = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
  return {
    ...(data as object),
    id,
    fromTransactionId: (data.fromTransactionId as string) ?? null,
    toTransactionId: (data.toTransactionId as string) ?? null,
    movementKind: (data.movementKind as ItemLineageMovementKind) ?? 'association',
    source: (data.source as ItemLineageSource) ?? 'server',
  } as ItemLineageEdge;
}

// ---------------------------------------------------------------------------
// Subscriptions
// ---------------------------------------------------------------------------

export function subscribeToEdgesFromTransaction(
  accountId: string,
  transactionId: string,
  onChange: (edges: ItemLineageEdge[]) => void,
): () => void {
  if (!isFirebaseConfigured || !db) {
    onChange([]);
    return () => {};
  }

  const q = query(
    collection(db, `accounts/${accountId}/lineageEdges`),
    where('fromTransactionId', '==', transactionId),
  );

  // Cache-first prelude: React Native Firebase's onSnapshot does NOT read
  // from cache on the first callback — it waits for a server round-trip.
  // getDocsFromCache gives us instant results while the listener spins up.
  getDocsFromCache(q)
    .then(snapshot => {
      const edges = snapshot.docs.map(d => normalizeEdgeFromFirestore(d.data(), d.id));
      onChange(edges);
    })
    .catch(() => {
      onChange([]);
    });

  // Real-time listener for ongoing updates
  return onSnapshot(
    q,
    (snapshot) => {
      const edges = snapshot.docs.map(d => normalizeEdgeFromFirestore(d.data(), d.id));
      onChange(edges);
    },
    (error) => {
      console.warn('[lineageEdgesService] subscription failed', error);
      onChange([]);
    },
  );
}

// ---------------------------------------------------------------------------
// Writes
// ---------------------------------------------------------------------------

export function createLineageEdge(
  accountId: string,
  edge: Omit<ItemLineageEdge, 'id' | 'createdAt'>,
): string {
  if (!isFirebaseConfigured || !db) {
    throw new Error(
      'Firebase is not configured. Add google-services.json / GoogleService-Info.plist and rebuild the dev client.',
    );
  }

  const edgeId = `app_${Date.now()}_${edge.itemId}_${edge.movementKind}`;
  const docRef = doc(collection(db, `accounts/${accountId}/lineageEdges`), edgeId);

  setDoc(docRef, {
    ...edge,
    createdAt: serverTimestamp(),
  }).catch(err => console.error('[lineageEdgesService] createLineageEdge failed:', err));

  trackPendingWrite();

  return edgeId;
}

File: LedgeriOS/LedgeriOS/Services/LineageEdgesService.swift
import FirebaseFirestore

struct LineageEdge: Codable, Identifiable {
    @DocumentID var id: String?
    var accountId: String?
    var itemId: String?
    var fromTransactionId: String?
    var toTransactionId: String?
    var movementKind: String?   // "sold" | "returned" | "correction" | "association"
    var source: String?         // "app" | "server" | "migration"
    var fromProjectId: String?
    var toProjectId: String?
    var note: String?
    var createdBy: String?

    @ServerTimestamp var createdAt: Date?
}

struct LineageEdgesService {
    private let db = Firestore.firestore()

    private func collection(accountId: String) -> CollectionReference {
        db.collection("accounts/\(accountId)/lineageEdges")
    }

    /// Fetches edges where the given transaction is either the source or destination.
    func edges(forTransaction transactionId: String, accountId: String) async throws -> [LineageEdge] {
        let col = collection(accountId: accountId)
        async let fromEdges = col
            .whereField("fromTransactionId", isEqualTo: transactionId)
            .getDocuments()
        async let toEdges = col
            .whereField("toTransactionId", isEqualTo: transactionId)
            .getDocuments()

        let (fromSnapshot, toSnapshot) = try await (fromEdges, toEdges)
        let fromParsed = fromSnapshot.documents.compactMap { try? $0.data(as: LineageEdge.self) }
        let toParsed = toSnapshot.documents.compactMap { try? $0.data(as: LineageEdge.self) }

        var seen = Set<String>()
        return (fromParsed + toParsed).filter { edge in
            guard let id = edge.id else { return false }
            return seen.insert(id).inserted
        }
    }

    /// Creates a new lineage edge document.
    func createEdge(_ edge: LineageEdge, accountId: String) throws {
        let ref = collection(accountId: accountId).document()
        try ref.setData(from: edge)
    }
}
